### PR TITLE
Refactored buildSlots

### LINKED ISSUE #6386: [CAL-745] Working Hours that are out of sync cause duplicate slots
Event Type has 1 hour intervals
User 1 starts the day at 9:00
User 2 starts the day at 9:15

image

The slots are duplicated, which user 1 rendered before user 2. On a COLLECTIVE event the following happens:

image

(Both users are able to attend the 9:15 slot, but as 9:00 is not available for User 2 all ranges on the whole hour are excluded - rendering stays consistent)

<sub>[CAL-745](https://linear.app/calcom/issue/CAL-745/working-hours-that-are-out-of-sync-cause-duplicate-slots)</sub>

### PR DESCRIPTION
**Pull request by emrysal**
## What does this PR do?

* 2-5x performance improvement on buildSlots, on a typical run taking ~220ms on a round-robin event with two users and regular working hours, 1h length, 15m intervals. On Vercel platform this is not running on M1 and the performance increase should be more (if 200ms isn't good enough).
* Booking Limit fixes

Fixes #6386

### The following was a round-robin event with 5 minute intervals, which is a bit skewed in benchmark favour. 

Before:
```
@calcom/web:dev: 261.1686658859253
@calcom/web:dev: 295.5547499656677
@calcom/web:dev: 276.91245889663696
@calcom/web:dev: 13.197291851043701
@calcom/web:dev: 201.46195888519287
@calcom/web:dev: 304.8555841445923
@calcom/web:dev: 348.87799978256226
@calcom/web:dev: 320.29675006866455
@calcom/web:dev: 298.3960838317871
@calcom/web:dev: 16.514708995819092
@calcom/web:dev: 412.25962495803833
@calcom/web:dev: 319.352792263031
@calcom/web:dev: 307.2695827484131
@calcom/web:dev: 331.67387533187866
@calcom/web:dev: 306.4797501564026
@calcom/web:dev: 323.24720764160156
@calcom/web:dev: 309.87562465667725
@calcom/web:dev: 335.6032090187073
@calcom/web:dev: 306.00012493133545
@calcom/web:dev: 322.82383394241333
@calcom/web:dev: 338.0093340873718
@calcom/web:dev: 293.10145902633667
@calcom/web:dev: 415.95374965667725
@calcom/web:dev: 278.45608282089233
@calcom/web:dev: 203.7317090034485
```
After:
```
@calcom/web:dev: 46.339165687561035
@calcom/web:dev: 42.44575023651123
@calcom/web:dev: 72.47133302688599
@calcom/web:dev: 58.1577091217041
@calcom/web:dev: 25.62195920944214
@calcom/web:dev: 49.89975023269653
@calcom/web:dev: 55.85516691207886
@calcom/web:dev: 51.87620830535889
@calcom/web:dev: 53.79541730880737
@calcom/web:dev: 54.397207736968994
@calcom/web:dev: 51.23254203796387
@calcom/web:dev: 52.09191608428955
@calcom/web:dev: 49.00404214859009
@calcom/web:dev: 52.425209045410156
```
### A regular event type still sees a significant performance improvement.

Before:
```
@calcom/web:dev: 111.88558435440063
@calcom/web:dev: 92.3113751411438
@calcom/web:dev: 113.30283308029175
@calcom/web:dev: 95.14154195785522
@calcom/web:dev: 117.83037519454956
@calcom/web:dev: 111.24895763397217
@calcom/web:dev: 107.13224983215332
@calcom/web:dev: 117.09833288192749
@calcom/web:dev: 109.87366724014282
@calcom/web:dev: 111.58466625213623
@calcom/web:dev: 98.55312490463257
@calcom/web:dev: 117.97745895385742
@calcom/web:dev: 106.8604998588562
@calcom/web:dev: 9.653749942779541
@calcom/web:dev: 113.2914171218872
@calcom/web:dev: 127.38441610336304
@calcom/web:dev: 118.18262481689453
@calcom/web:dev: 117.22816705703735
@calcom/web:dev: 108.8679575920105
@calcom/web:dev: 136.15558290481567
@calcom/web:dev: 155.59845876693726
@calcom/web:dev: 134.73008394241333
@calcom/web:dev: 138.19720792770386
@calcom/web:dev: 129.73195791244507
@calcom/web:dev: 118.53083276748657
```
After:
```
@calcom/web:dev: 77.42591619491577
@calcom/web:dev: 68.74491691589355
@calcom/web:dev: 73.46024990081787
@calcom/web:dev: 74.83408403396606
@calcom/web:dev: 60.6900839805603
@calcom/web:dev: 48.18441581726074
@calcom/web:dev: 67.67670917510986
@calcom/web:dev: 61.3255410194397
@calcom/web:dev: 81.16716623306274
@calcom/web:dev: 56.736083984375
@calcom/web:dev: 51.41737508773804
@calcom/web:dev: 3.710665702819824
@calcom/web:dev: 44.120041847229004
@calcom/web:dev: 2.113999843597412
@calcom/web:dev: 49.696624755859375
@calcom/web:dev: 54.930707931518555
@calcom/web:dev: 48.09649991989136
@calcom/web:dev: 1.8699584007263184
@calcom/web:dev: 51.06704235076904
@calcom/web:dev: 2.062624931335449
@calcom/web:dev: 42.9293327331543
@calcom/web:dev: 49.08449983596802
@calcom/web:dev: 1.9262089729309082
@calcom/web:dev: 44.59933280944824
@calcom/web:dev: 47.761208057403564
```

### CODE REVIEW

File: packages/lib/slots.ts [hunk 1]
```diff
@@ -15,41 +15,6 @@ export type TimeFrame = { userIds?: number[]; startTime: number; endTime: number
 
 const minimumOfOne = (input: number) => (input < 1 ? 1 : input);
 
-/**
- * TODO: What does this function do?
- * Why is it needed?
```
Comment (emrysal) [inline packages/lib/slots.ts:20]: @zomars ;)

File: packages/trpc/server/routers/viewer/slots.tsx [hunk 2]
```diff
@@ -280,85 +279,93 @@ export async function getSchedule(input: z.infer<typeof getScheduleSchema>, ctx:
...
-                  time: slot.time,
-                  ...userSchedule,
-                  ...availabilityCheckProps,
-                });
-              });
-            })()
-      );
+  const timeSlots: ReturnType<typeof getTimeSlots> = [];
+
+  for (
```
Comment (emrysal) [inline packages/trpc/server/routers/viewer/slots.tsx:289]: `do {} while {}` changed in a `for` loop, flattened into one big array for decreasing depth.

File: packages/core/getUserAvailability.ts [hunk 3]
```diff
@@ -164,7 +164,7 @@ export async function getUserAvailability(
     let startDate = dayjs(dateFrom);
     const endDate = dayjs(dateTo);
     while (startDate.isBefore(endDate)) {
-      dates.push(startDate.add(1, "day"));
+      dates.push(startDate);
```
Comment (emrysal) [inline packages/core/getUserAvailability.ts:167]: This added a day to all dates, which primarily showed itself by breaking the day booking limit but probably had knock-on effects with week/month/year.

**Comment by emrysal** [comment]
@hariombalhara I've got one failing test which I believe shouldn't be failing - I tested Google Calendar on my dev and it works there, also the test is passing on my local machine. Excellent tests by the way as they caught an early cut-off. Could you have a look at the last test let me know your thoughts?

It's also strange because this PR doesn't really touch busy time logic.

**Comment by hariombalhara** [comment]
@emrysal I found 2 things:
1. 12:15 timeslot is being returned now which wasn't returned earlier. I am not sure if it's expected. If availability is till 12:30, does it mean that 12:15 timeslot(of 45mins) can be booked which would actually require the user to be available for meeting till 1PM.
2. Calendar busy check is failing when all tests run together. This is because jest tests run in parallel and I haven't properly handled nock with parallel tests. I am not sure why they didn't fail earlier or not failing even now. The failure is only in your PR.

File: apps/web/test/lib/getSchedule.test.ts [hunk 4]
```diff
@@ -785,6 +777,7 @@ describe("getSchedule", () => {
           {
             id: 1,
             slotInterval: 45,
+            length: 45,
```
Comment (hariombalhara) [inline apps/web/test/lib/getSchedule.test.ts:780]: If it's possible to have `length` unset in production, we should have it the same way.
Maybe we can create both types of tests(where `length` is set and where it's not set)

**Review by zomars — APPROVED** [review]
So good! LFG!! LGTM

**Pull request merged**